api/src/main/java/org/openmrs/module/appframework/service/AppFrameworkServiceImpl.java

```diff
@@ -356,6 +369,62 @@ public boolean checkRequireExpression(Extension candidate, AppContextModel conte
         }
     }
 
+    protected String getPatientUuid(AppContextModel contextModel) {
+    	Bindings bindings = new SimpleBindings();
+	    for (Map.Entry<String, Object> e : contextModel.entrySet()) {
+             bindings.put(e.getKey(), e.getValue());
+	    }
+        javascriptEngine.setBindings(bindings, ScriptContext.ENGINE_SCOPE);
+    	try {
+			String uuid = (String) javascriptEngine.eval("patient.uuid;");
+			if (StringUtils.isBlank(uuid)) {
+	    		throw new APIException("Patient uuid cannot be empty");
+			}
+			return uuid;
+		} catch (ScriptException e) {
+    		    	throw new APIException("Failed to get patient uuid", e);
+		}
+    }
+    
+    protected boolean checkRequiredState(Extension candidate, AppContextModel contextModel) {
+    	String requiredState = candidate.getRequiredState();
+    	if (StringUtils.isBlank(requiredState)) {
+    		return true;
+    	}
+    	Concept stateConcept = AppFrameworkUtil.getConcept(requiredState);
```

Comment:
You should get the `ProgramWorflowState` instance from this concept, that's what you really want to use when "requiring a state".

Reply:
@samuelmale you didn't come back to me on this one and its consequence here: https://github.com/openmrs/openmrs-module-appframework/pull/40#discussion_r317101205

Reply:
Well this makes sense to me, but the limiting factor is with the `ProgramWorkflowService`. It's very shallow and there is no straight way of getting a `ProgramWorflowState` using a concept. Unless looping through all available ProgramWorflows and loop through their ProgramWorflowStates and compare the concepts. Which if find so handy! But we can still use the concept; I mean it works! I had no choice anyway!